### ISSUE TITLE
Can you make an alternative, pure python version of Tiktoken? (no Rust dependency)

### ISSUE THREAD

**Issue by Emasoft** [openai/tiktoken#36]
Can you make an alternative, pure python version of Tiktoken, with no Rust dependency?
For those who cannot compile and run Rust binaries on their system (for various reasons: package managers support, company policy, intranet or local machine security, docking containers limitations, vm restrictions, environment virtualization, lack of Rust support in jupyter notebooks remote hosting, etc).
Of course it would be slower than the Rust version, but as a fallback it would allow Tiktoken to be used in a much wider set of platforms and working environments.
Thanks!

**Comment by Emasoft** [comment]
@hauntsaninja ?

**Comment by hauntsaninja** [comment]
I have no current plans to support this.

**Comment by hauntsaninja** [comment]
I'm closing the issue as per^. Note that the latest release of tiktoken includes aarch64 wheels, in case that was the issue you were running into.

**Comment by kechan** [comment]
@Emasoft I want to actually run tiktoken in code interpreter but it isn't available. so I had to tar up the git repos and run it by adding it to sys path. This approach won't work due to the rust binary. And it looks like one of case you need a pure python impl. I don't know RUST but GPT4 did a pretty good job helping me translating it (plus a few minor debugging and re-trials), you can consult: 

https://github.com/kechan/tiktoken.git

I added the option to Encoding class, and you will have to instantiate it a different way to invoke the pure python tokenizing. Or you can hard code and hack it up. 

I tested quite a few symbols and sentences, and it seems ok. (GPT4 is awesome).

**Comment by Emasoft** [comment]
@kechan This is great! Thank you! I'll try this asap.

**Comment by kechan** [comment]
@Emasoft I used it like this locally (but there may be easier/better way):

import tiktoken.registry as registry
from tiktoken.registry import _find_constructors
from tiktoken.core import Encoding

_find_constructors()
constructor = registry.ENCODING_CONSTRUCTORS['cl100k_base']
params = constructor()
enc = Encoding(**params, use_pure_python=True)
enc.encode("hello world")

For code interpreter, I sort of pickle the params (a dict), upload it and then feed in **params like above, and able to call the encoder. it is quite handy for experimentation, so I didn't do exhaustive test. if you find any issue, let me know. Likely will enlist gpt4's RUST skill to fix.

**Comment by MRYingLEE** [comment]
@kechan Can you publish the pure Python wheel in your repo? I struggled to build it.

**Comment by kechan** [comment]
@MRYingLEE Instead of trying to build or pip install, just tar up the tiktoken repository, untar it on the target machine (with limited ability or restriction), and use 

sys.path.append(os.path.join(extract_path, 'tiktoken'))

to add it to your path. On the more capable machine with normal install, run the code I posted to get & pickle the params for cl100k_base, and use that same file on the target machine. 

in short, I don't have time to separate out the rust dependency from setup.py and build process, I just use sys.path.append the location where I untar the repos. maybe I will come around later. Hope this helps.

**Comment by hauntsaninja** [comment]
Note that the educational submodule in tiktoken contains pure Python code for using and training BPE encodings. https://github.com/openai/tiktoken/blob/main/tiktoken/_educational.py

It has an underscore prefix since I don't want to promise any API stability, but could be useful for folks finding this issue

**Comment by KillianLucas** [comment]
Must be a simpler way to do this? Count tokens without rust?

**Comment by Emasoft** [comment]
> Note that the educational submodule in tiktoken contains pure Python code for using and training BPE encodings. https://github.com/openai/tiktoken/blob/main/tiktoken/_educational.py

@hauntsaninja  Since the pure python code version it's already available, why do not add to the pip setup the option to install the pure python version without Rust dependencies on platforms without Rust installed? More and more people are in need of such version, for a variety of reasons. Please! 🙏
